lib/lightning/digest_email_worker.ex

```diff
@@ -43,17 +50,116 @@ defmodule Lightning.DigestEmailWorker do
       digest_data =
         Workflows.get_workflows_for(pu.project)
         |> Enum.map(fn workflow ->
-          Workorders.get_digest_data(workflow, digest)
+          get_digest_data(workflow, start_date, end_date)
         end)
 
       UserNotifier.deliver_project_digest(
-        pu.user,
-        pu.project,
         digest_data,
-        digest
+        %{
+          user: pu.user,
+          project: pu.project,
+          digest: digest,
+          start_date: start_date,
+          end_date: end_date
+        }
       )
     end)
 
-    {:ok, %{project_users: project_users}}
+    {:ok, project_users}
   end
+
+  def digest_to_date(digest) do
+    case digest do
+      :daily ->
+        Timex.now() |> Timex.beginning_of_day()
+
+      :weekly ->
+        Timex.now() |> Timex.shift(days: -7) |> Timex.beginning_of_week()
+
+      :monthly ->
+        Timex.now() |> Timex.shift(months: -1) |> Timex.beginning_of_month()
+    end
+  end
+
+  @doc """
+  Get a map of counts for successful, rerun and failed Workorders for a given workflow in a given timeframe.
+  """
+  def get_digest_data(workflow, start_date, end_date) do
+    successful_workorders = successful_workorders(workflow, start_date, end_date)
+    failed_workorders = failed_workorders(workflow, start_date, end_date)
+
+    %{
+      workflow: workflow,
+      successful_workorders: successful_workorders.total_entries,
+      failed_workorders: failed_workorders.total_entries
+    }
+  end
+
+  defp successful_workorders(workflow, start_date, end_date) do
+    Lightning.Invocation.list_work_orders_for_project(
+      Projects.get_project!(workflow.project_id),
+      [
+        status: [:success],
+        search_fields: [:body, :log],
+        search_term: "",
+        workflow_id: workflow.id,
+        date_after: start_date,
+        date_before: end_date,
+        wo_date_after: "",
+        wo_date_before: ""
+      ],
+      %{}
+    )
+  end
+
+  defp failed_workorders(workflow, start_date, end_date) do
+    Lightning.Invocation.list_work_orders_for_project(
+      Projects.get_project!(workflow.project_id),
+      [
+        status: [:failure, :crash, :timeout],
+        search_fields: [:body, :log],
+        search_term: "",
+        workflow_id: workflow.id,
+        date_after: start_date,
+        date_before: end_date,
+        wo_date_after: "",
+        wo_date_before: ""
+      ],
+      %{}
+    )
```

Comment:
This interface feels a little off to me namely:

1. Things like `search_term` being an empty string, is that a default or an explicit search for `""` (likely not the case). Could these empty values be considered defaults and if they can, would we be able to omit them?
2. The 3 arguments (the 2nd being the filter, and the 3rd being pagination params), I think the filter should be changed to a map, and the params be defaulted to an empty map (`\\ %{}`).

Perhaps it's a good moment to create a struct for the search params (if needed), something that can hold defaults and coercion to keywords or uri params etc.

Something like:

```elixir
defmodule WorkorderSearch do
  defstruct [:status, :search_fields, ...]

  def new(attrs) do
    // possibly merge in defaults
    struct!(__MODULE__, attrs)
  end

  def to_params(search) do
    Map.from_struct(search) // possibly strip empty/null values?
    |> stringify_keys()
  end
end
```

This kind of approach would be used to replace part of `build_digest_url` and ensure the building up of search params stays in one place

Reply:
Done